### PR TITLE
Put clothing on people by clicking on them with clothing inhand + targeting correct area.

### PR DESCRIPTION
**Pull request by itseasytosee**
## About The Pull Request
Adds a element that when added to an item will make the human attempt to dress an attacked target with the item while not in combat mode provided they are targeting the right slot. This element is attached to all clothing, backpacks and belts.

HEAD - Ears, Helmet, Neck
EYES - Glasses
MOUTH - Mask
CHEST - Jumpsuit, Outer suit, Backpack
ARMS - Gloves
GROIN - Belt
LEGS - Shoes.

Demonstration vid:
https://streamable.com/2ovika

This is my first time dealing with elements so tell me if I fucked up.

## Why It's Good For The Game
Its pretty cool, feels nice.

## Changelog

:cl: itseasytosee
qol: You can now put clothing, backpacks, and belts onto people by clicking on them with it in your hand and targeting the right area.
/:cl:

### CODE REVIEW

File: code/datums/elements/strippable.dm [hunk 1]
```diff
@@ -447,7 +447,7 @@
 						return
 
 					// They equipped an item in the meantime
-					if (!isnull(strippable_item.get_item(owner)))
+					if(!isnull(strippable_item.get_item(owner)))
```
Comment (Mothblocks) [inline code/datums/elements/strippable.dm:450]: don't touch this file if you're not doing anything with it other than this

File: code/datums/elements/attack_equip.dm [hunk 2]
```diff
@@ -0,0 +1,72 @@
+/datum/element/attack_equip
```
Comment (Mothblocks) [inline code/datums/elements/attack_equip.dm:1]: document

File: code/datums/elements/attack_equip.dm [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+/datum/element/attack_equip/Attach(datum/target)
+	. = ..()
+	if(!isitem(target))
+		return ELEMENT_INCOMPATIBLE
+	RegisterSignal(target, COMSIG_ITEM_ATTACK, .proc/onItemAttack)
```
Comment (Mothblocks) [inline code/datums/elements/attack_equip.dm:7]: on_item_attack

File: code/modules/clothing/clothing.dm [hunk 4]
```diff
@@ -106,17 +107,16 @@
...
 		return ..()
 	if(clothing_flags & INEDIBLE_CLOTHING)
 		return ..()
 	if(isnull(moth_snack))
 		moth_snack = new
 		moth_snack.name = name
 		moth_snack.clothing = WEAKREF(src)
-	moth_snack.attack(M, user, params)
-
+	moth_snack.attack(target, user, params)
```
Comment (Mothblocks) [inline code/modules/clothing/clothing.dm:119]:
```suggestion
	moth_snack.attack(target, user, params)

```

**Comment by exymian** [comment]
As convenient as this is, you have to think about the combat aspect. This could be used to remove armor or steal clothing.

**Comment by Mothblocks** [comment]
i dont understand. this only equips

File: code/datums/elements/strippable.dm (hunk 1 above)
Comment (itseasytosee) [reply code/datums/elements/strippable.dm:450]: Yep, my mistake

**Comment by itseasytosee** [comment]
> As convenient as this is, you have to think about the combat aspect. This could be used to remove armor or steal clothing.

It indeed, could not.

**Comment by VladinXXV** [comment]
Make sure this doesn't make it faster to put a straightjacket on? I'm also concerned with weapons accidentally attempting to be equipped to someone, such as nullrods, which often take up the belt/back slot. I tried to read the code the best I could and I think these are possibilities?

**Comment by itseasytosee** [comment]
> Make sure this doesn't make it faster to put a straightjacket on? I'm also concerned with weapons accidentally attempting to be equipped to someone, such as nullrods, which often take up the belt/back slot. I tried to read the code the best I could and I think these are possibilities?

Nope, this component is only applied on subtypes of clothing, backpacks and belts. I can't think of a single weapon that falls into this category, and if there were you could still attack them with combat mode.

It takes the same time to put on a straight jacket as normal, being 2 seconds, unless you count time spent opening the strip menu. If you thing this is not enough time, then that already applies to putting on the straight jacket normally. 

If you want to peruse a balance change of this nature, I would recommend going to the straightjacket.dm file, and defining the variable equip_delay_other as a greater number. Just as equip_delay_self is defined as 50. By default the number you put in here is a measurement is 
deciseconds, but its good practice to write it in seconds and the writing SECONDS (in all caps) following it, which essentially multiplies the number by ten and is easier to read.

**Comment by SuperNovaa41** [comment]
Bring back shoe stealing

**Comment by itseasytosee** [comment]
> Bring back shoe stealing

Now that we have combat mode we actually could

**Comment by Mothblocks** [comment]
out of scope and also it needs to be an emote like *slap

File: code/game/objects/items.dm [hunk 5]
```diff
@@ -1480,3 +1480,33 @@ GLOBAL_DATUM_INIT(welding_sparks, /mutable_appearance, mutable_appearance('icons
...
+/**
+ *  When called on an item, and given a body targeting zone, this will return TRUE if the item slot matches the target zone, and FALSE otherwise.
+ *
+ * Currently supports the jumpsuit, outersuit, backpack, belt, gloves, hat, ears, neck, mask, eyes, and feet slots. All other slots will auto return FALSE.
+ */
+/obj/item/proc/compare_zone_to_item_slot(zone)
+	switch(slot_flags)
+		if(ITEM_SLOT_ICLOTHING, ITEM_SLOT_OCLOTHING, ITEM_SLOT_BACK)
+			if(zone == BODY_ZONE_CHEST)
+				return TRUE
```
Comment (Mothblocks) [inline code/game/objects/items.dm:1493]: change these all to just `return zone == BODY_ZONE_CHEST` etc

File: code/datums/elements/attack_equip.dm [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+/datum/element/attack_equip/proc/start_equip(obj/item/attire, mob/living/carbon/human/sharp_dresser, mob/living/user)
+
+	if(HAS_TRAIT(attire, TRAIT_NODROP))
+		to_chat(user, span_warning("You can't put [attire] on [sharp_dresser], it's stuck to your hand!"))
+		return
+	var/equip_time = attire.equip_delay_other
+	if(isclothing(attire))
+		var/obj/item/clothing/fancy_tux = attire
+		if(fancy_tux.clothing_flags == DANGEROUS_OBJECT) // give us danger text if we have a dangerous object on our hands
```
Comment (Mothblocks) [inline code/datums/elements/attack_equip.dm:43]: i am not happy about this code being duplicated from here and normal equip code. can you not share them?

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/elements/attack_equip.dm (hunk 6 above)
Comment (itseasytosee) [reply code/datums/elements/attack_equip.dm:43]: I suppose I could. My main Idea for implementing that would make it a proc on items that both this and the strip menu call, would that work?

File: code/game/objects/items.dm (hunk 5 above)
Comment (itseasytosee) [reply code/game/objects/items.dm:1493]: Roger

File: code/datums/elements/attack_equip.dm (hunk 6 above)
Comment (Mothblocks) [reply code/datums/elements/attack_equip.dm:43]: yes

**Comment by itseasytosee** [comment]
👍

**Comment by JohnFulpWillard** [comment]
Maybe medals can use this system too? That way it would also add it for pocket protectors (and by extension, detective's camera bug)

**Comment by itseasytosee** [comment]
> Maybe medals can use this system too? That way it would also add it for pocket protectors (and by extension, detective's camera bug)

Its not compatible as is, because accessories don't fit into clothing slots (most of the time), they attach to clothing

File: code/datums/elements/attack_equip.dm [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		return COMPONENT_CANCEL_ATTACK_CHAIN
+
+
+/datum/element/attack_equip/proc/equip(obj/item/attire, mob/living/carbon/human/sharp_dresser, mob/living/user)
+
+	if(HAS_TRAIT(attire, TRAIT_NODROP))
+		to_chat(user, span_warning("You can't put [attire] on [sharp_dresser], it's stuck to your hand!"))
+		return
+	var/equip_time = attire.equip_delay_other
+	if(isclothing(attire))
```
Comment (JohnFulpWillard) [inline code/datums/elements/attack_equip.dm:41]: are you gonna make this add the new generalized code you added?

File: code/game/objects/items.dm [hunk 8]
```diff
@@ -1503,3 +1503,36 @@ GLOBAL_DATUM_INIT(welding_sparks, /mutable_appearance, mutable_appearance('icons
 		if(ITEM_SLOT_FEET)
 			return (zone == BODY_ZONE_L_LEG || zone == BODY_ZONE_R_LEG)
 	return FALSE
+
+/obj/item/proc/item_start_equip(atom/target, obj/item/equipping, mob/user, warn_dangerous = TRUE)
```
Comment (Mothblocks) [inline code/game/objects/items.dm:1507]: Document it

**Comment by itseasytosee** [comment]
👍

File: code/datums/elements/attack_equip.dm [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+	if(attire.mob_can_equip(target, user,  attire.slot_flags, disable_warning = TRUE, bypass_equip_delay_self = TRUE))
+		INVOKE_ASYNC(src, .proc/equip, attire, sharp_dresser, user)
+		return COMPONENT_CANCEL_ATTACK_CHAIN
+
+
+/datum/element/attack_equip/proc/equip(obj/item/attire, mob/living/carbon/human/sharp_dresser, mob/living/user)
+
+	if(HAS_TRAIT(attire, TRAIT_NODROP))
+		to_chat(user, span_warning("You can't put [attire] on [sharp_dresser], it's stuck to your hand!"))
+		return
```
Comment (JohnFulpWillard) [inline code/datums/elements/attack_equip.dm:39]: Is TRAIT_NODROP bypassing ``mob_can_equip``?

Comment (itseasytosee) [reply code/datums/elements/attack_equip.dm:39]: mob_can_equip is checked on_item_attack() and in the strippable component. So I don't believe its an issue.

**Pull request merged**